In [17]:
import pandas as pd
import collections
import pickle

# Загрузка данных по коду с ошибками

Загрузим данные всех датасетов

In [3]:
source_code_data_train = pd.read_pickle("filtered_traceback_bugfixes_train.pickle")
source_code_data_valid = pd.read_pickle("filtered_traceback_bugfixes_valid.pickle")
source_code_data_test = pd.read_pickle("bugfixes_test.pickle")

Создадим колонку с названием репозитория, содержащего каждый сниппет.

In [4]:
source_code_data_train['repo_name'] = source_code_data_train['url'].apply(lambda x: '/'.join(x.split('/')[3:5]))
source_code_data_valid['repo_name'] = source_code_data_valid['url'].apply(lambda x: '/'.join(x.split('/')[3:5]))
source_code_data_test['repo_name'] = source_code_data_test['url'].apply(lambda x: '/'.join(x.split('/')[3:5]))

In [5]:
def load_repo_names(dataset):
    return list(set(dataset["repo_name"].to_list()))

In [6]:
train_repos = load_repo_names(source_code_data_train)
valid_repos = load_repo_names(source_code_data_valid)
test_repos = load_repo_names(source_code_data_test)

Посмотрим сколько репозиториев представлено в коде с ошибками.

In [8]:
len(test_repos), len(train_repos), len(valid_repos), len(test_repos) + len(train_repos) + len(valid_repos)

(68, 562, 68, 698)

Посмотрим сколько сниппетов лежит в каждом репозитории

In [115]:
train_repos_count = source_code_data_train['repo_name'].value_counts()

for i, count in zip(train_repos_count.index, train_repos_count):
    print(i, count)

pandas-dev/pandas 709
mars-project/mars 635
numba/numba 607
python/mypy 342
conda/conda 314
ray-project/ray 261
pytroll/satpy 229
iterative/dvc 221
modin-project/modin 220
qutebrowser/qutebrowser 200
google/jax 184
MobSF/Mobile-Security-Framework-MobSF 135
DataBiosphere/toil 126
galaxyproject/galaxy 123
OctoPrint/OctoPrint 119
janeczku/calibre-web 116
riptideio/pymodbus 113
cornellius-gp/gpytorch 108
matrix-org/synapse 107
numpy/numpy 104
napari/napari 104
pantsbuild/pex 102
leo-editor/leo-editor 96
statsmodels/statsmodels 92
sqlalchemy/sqlalchemy 92
conan-io/conan 91
nucypher/nucypher 91
Trusted-AI/adversarial-robustness-toolbox 84
WeblateOrg/weblate 83
hbldh/bleak 82
cloud-custodian/cloud-custodian 82
edgedb/edgedb 79
google/pytype 77
celery/celery 72
hyperspy/hyperspy 71
yt-project/yt 70
tox-dev/tox 70
pantsbuild/pants 69
ipython/ipython 68
beetbox/beets 66
mopidy/mopidy 66
RaRe-Technologies/gensim 66
spack/spack 66
mne-tools/mne-python 65
snakemake/snakemake 64
hylang/hy 64
apache/

In [117]:
valid_repos_count = source_code_data_valid['repo_name'].value_counts()

for i, count in zip(valid_repos_count.index, valid_repos_count):
    print(i, count)

saltstack/salt 1801
ansible/ansible 1406
Tribler/tribler 717
spyder-ide/spyder 551
Cog-Creators/Red-DiscordBot 440
pymedusa/Medusa 352
log2timeline/plaso 307
scikit-learn/scikit-learn 244
freqtrade/freqtrade 230
sphinx-doc/sphinx 216
scipy/scipy 201
PyTorchLightning/pytorch-lightning 162
Azure/azure-cli 159
pydata/xarray 146
mitmproxy/mitmproxy 140
CellProfiler/CellProfiler 138
RasaHQ/rasa 129
getnikola/nikola 129
xonsh/xonsh 119
custom-components/alexa_media_player 98
Flexget/Flexget 97
nicolargo/glances 93
pypa/pipenv 91
Qiskit/qiskit-terra 90
certtools/intelmq 83
pypa/pip 81
spesmilo/electrum 77
ParallelSSH/parallel-ssh 68
biolab/orange3 64
databricks/koalas 59
holoviz/holoviews 58
pydicom/pydicom 57
evennia/evennia 55
awslabs/gluon-ts 53
facebookresearch/hydra 50
psychopy/psychopy 48
zulip/zulip 45
frappe/erpnext 38
SciTools/iris 36
quodlibet/quodlibet 35
certbot/certbot 35
fossasia/open-event-server 31
ibis-project/ibis 28
aiogram/aiogram 28
postlund/pyatv 27
DMOJ/judge-server 26


In [118]:
test_repos_count = source_code_data_test['repo_name'].value_counts()

for i, count in zip(test_repos_count.index, test_repos_count):
    print(i, count)

certtools/intelmq 17
ansible/ansible 14
scikit-image/scikit-image 10
Rapptz/discord.py 9
frappe/erpnext 8
saltstack/salt 7
xonsh/xonsh 5
espnet/espnet 5
pypa/pipenv 4
Tribler/tribler 4
mitmproxy/mitmproxy 3
PyTorchLightning/pytorch-lightning 3
SciTools/iris 3
spyder-ide/spyder 2
freqtrade/freqtrade 2
Cog-Creators/Red-DiscordBot 2
python-poetry/poetry 2
pymedusa/Medusa 2
scikit-learn/scikit-learn 2
Azure/azure-cli 2
jazzband/django-debug-toolbar 2
scipy/scipy 2
Flexget/Flexget 2
jpadilla/pyjwt 2
evennia/evennia 2
biolab/orange3 2
sphinx-doc/sphinx 2
Qiskit/qiskit-terra 1
spesmilo/electrum 1
OpenShot/openshot-qt 1
aiogram/aiogram 1
psychopy/psychopy 1
typeddjango/django-stubs 1
nicolargo/glances 1
google/mobly 1
PrefectHQ/prefect 1
ansible-community/molecule 1
holoviz/holoviews 1
awslabs/gluon-ts 1
certbot/certbot 1
databricks/koalas 1
GenericMappingTools/pygmt 1
astropy/astroquery 1
pypa/pip 1
zulip/zulip 1
pydicom/pydicom 1
pyocd/pyOCD 1
getnikola/nikola 1
log2timeline/plaso 1
Parallel

Сниппеты обучающей и тестовой выборок не пересекаются по репозиториям:

In [13]:
set(test_repos) & set(train_repos)

set()

Репозитории в валидационной и тестовой выборках совпадают:

In [14]:
set(test_repos) == set(valid_repos)

True

# Топики репозиториев

Загрузим данные с топиками репозиториев для кода с ошибками.

In [62]:
repo_name_to_topics = pickle.load(open("topics_slashes.pickle", "rb"))
train_repos_topics = [repo_name_to_topics[repo_name] for repo_name in source_code_data_train['repo_name'].to_numpy() if repo_name in repo_name_to_topics]
valid_repos_topics = [repo_name_to_topics[repo_name] for repo_name in source_code_data_valid['repo_name'].to_numpy()  if repo_name in repo_name_to_topics]
test_repos_topics = [repo_name_to_topics[repo_name] for repo_name in source_code_data_test['repo_name'].to_numpy()  if repo_name in repo_name_to_topics]

In [63]:
train_repos_topics_cnt = collections.Counter()
for t in train_repos_topics:
    train_repos_topics_cnt.update(set(t))
    
valid_repos_topics_cnt = collections.Counter()
for t in valid_repos_topics:
    valid_repos_topics_cnt.update(set(t))

test_repos_topics_cnt = collections.Counter()
for t in test_repos_topics:
    test_repos_topics_cnt.update(set(t))

Наиболее представленные топики в обучающей выборке:

In [68]:
train_repos_topics_cnt.most_common()

[('python', 9538),
 ('machine-learning', 2239),
 ('hacktoberfest', 2098),
 ('pandas', 1733),
 ('numpy', 1542),
 ('pytorch', 1378),
 ('ray', 1146),
 ('data-science', 986),
 ('deep-learning', 950),
 ('tensorflow', 947),
 ('parallel', 906),
 ('data-analysis', 876),
 ('dataframe', 864),
 ('scikit-learn', 755),
 ('flexible', 736),
 ('alignment', 709),
 ('tensor', 686),
 ('compiler', 672),
 ('rocm', 658),
 ('cuda', 657),
 ('statsmodels', 635),
 ('lightgbm', 635),
 ('joblib', 635),
 ('xgboost', 635),
 ('llvm', 607),
 ('linter', 566),
 ('python3', 496),
 ('distributed', 490),
 ('ai', 443),
 ('typing', 419),
 ('typechecker', 419),
 ('types', 419),
 ('automl', 408),
 ('aws', 401),
 ('visualization', 381),
 ('django', 365),
 ('cli', 349),
 ('asyncio', 343),
 ('hyperparameter-optimization', 340),
 ('java', 332),
 ('sql', 329),
 ('workflow', 320),
 ('static-analysis', 315),
 ('package-management', 315),
 ('conda', 314),
 ('reproducibility', 307),
 ('linux', 304),
 ('deployment', 299),
 ('natural-la

Наиболее представленные топики в валидационной выборке:

In [69]:
valid_repos_topics_cnt.most_common()

[('python', 7162),
 ('hacktoberfest', 2365),
 ('cloud', 1960),
 ('cloud-provisioning', 1801),
 ('infrastructure-as-code', 1801),
 ('event-management', 1801),
 ('infrastructure-automation', 1801),
 ('infrastructure-as-a-code', 1801),
 ('zeromq', 1801),
 ('cloud-providers', 1801),
 ('event-stream', 1801),
 ('edge', 1801),
 ('remote-execution', 1801),
 ('iot', 1801),
 ('infrastructure-management', 1801),
 ('configuration-management', 1801),
 ('cloud-management', 1801),
 ('infrasructure', 1801),
 ('ansible', 1415),
 ('p2p', 717),
 ('privacy', 717),
 ('decentralized', 717),
 ('torrent-client', 717),
 ('bittorrent', 717),
 ('search', 717),
 ('streaming', 717),
 ('tor', 717),
 ('machine-learning', 661),
 ('bot', 597),
 ('chatbot', 569),
 ('winpython', 551),
 ('spyder', 551),
 ('anaconda-distribution', 551),
 ('anaconda', 551),
 ('language-server-protocol', 551),
 ('qt', 551),
 ('pyqt5', 551),
 ('data-science', 545),
 ('music', 475),
 ('discord', 465),
 ('tv', 449),
 ('torrent', 449),
 ('modul

Наиболее представленные топики в тестовой выборке:

In [70]:
test_repos_topics_cnt.most_common()

[('python', 131),
 ('hacktoberfest', 34),
 ('automation', 21),
 ('cert', 17),
 ('malware', 17),
 ('ioc', 17),
 ('csirt', 17),
 ('intelligence', 17),
 ('threat', 17),
 ('feeds', 17),
 ('ihap', 17),
 ('cybersecurity', 17),
 ('phishing', 17),
 ('incident-response', 17),
 ('alerts', 17),
 ('incident', 17),
 ('handling', 17),
 ('ansible', 15),
 ('discord', 11),
 ('deep-learning', 10),
 ('machine-learning', 10),
 ('computer-vision', 10),
 ('image-processing', 10),
 ('cloud', 9),
 ('data-science', 9),
 ('pytorch', 9),
 ('python-3', 9),
 ('discord-api', 9),
 ('erpnext', 8),
 ('healthcare', 8),
 ('distribution', 8),
 ('manufacturing', 8),
 ('integrations', 8),
 ('retail', 8),
 ('asset-management', 8),
 ('localization', 8),
 ('frappe', 8),
 ('profit', 8),
 ('hrms', 8),
 ('erp', 8),
 ('support', 8),
 ('accounting', 8),
 ('agriculture', 8),
 ('crm', 8),
 ('project-management', 8),
 ('point-of-sale', 8),
 ('procurement', 8),
 ('cloud-provisioning', 7),
 ('infrastructure-as-code', 7),
 ('event-manag

Сравним представленности топиков:

In [64]:
train_topics_count = pd.DataFrame(train_repos_topics_cnt.most_common())
train_topics_count.rename({0: 'topic name', 1: 'train count'},
                           inplace=True,
                           axis=1)
train_topics_count.set_index(keys='topic name', inplace=True)

valid_topics_count = pd.DataFrame(valid_repos_topics_cnt.most_common())
valid_topics_count.rename({0: 'topic name', 1: 'valid count'},
                           inplace=True,
                           axis=1)
valid_topics_count.set_index(keys='topic name', 
                             inplace=True)
test_topics_count = pd.DataFrame(test_repos_topics_cnt.most_common())
test_topics_count.rename({0: 'topic name', 1: 'test count'},
                           inplace=True,
                           axis=1)
test_topics_count.set_index(keys='topic name', inplace=True)

In [65]:
topics_count = pd.concat((train_topics_count, valid_topics_count, test_topics_count), 
                         axis=1)

In [66]:
topics_count['train topic ratio'] = topics_count['train count'] / source_code_data_train.shape[0]
topics_count['valid topic ratio'] = topics_count['valid count'] / source_code_data_valid.shape[0]
topics_count['test topic ratio'] = topics_count['test count'] / source_code_data_test.shape[0]

In [71]:
topics_count = topics_count.fillna(0)
topics_count

train count  valid count  test count  train topic ratio  \
topic name                                                                  
python                 9538.0       7162.0       131.0           0.675591   
machine-learning       2239.0        661.0        10.0           0.158592   
hacktoberfest          2098.0       2365.0        34.0           0.148605   
pandas                 1733.0        297.0         5.0           0.122751   
numpy                  1542.0        210.0         3.0           0.109222   
...                       ...          ...         ...                ...   
mobile                    0.0          7.0         1.0           0.000000   
android-devices           0.0          7.0         1.0           0.000000   
android-wear              0.0          7.0         1.0           0.000000   
jwt                       0.0          2.0         2.0           0.000000   
geoscience                0.0          1.0         1.0           0.000000   

                  valid topic ratio  test topic ratio  
topic name                                             
python                     0.757323          0.813665  
machine-learning           0.069895          0.062112  
hacktoberfest              0.250079          0.211180  
pandas                     0.031405          0.031056  
numpy                      0.022206          0.018634  
...                             ...               ...  
mobile                     0.000740          0.006211  
android-devices            0.000740          0.006211  
android-wear               0.000740          0.006211  
jwt                        0.000211          0.012422  
geoscience                 0.000106          0.006211  

[2573 rows x 6 columns]

Отсортируем по столбцу `train topic ratio`.

In [73]:
topics_count.sort_values(by='train topic ratio',
                         ascending=False).head(30)

train count  valid count  test count  train topic ratio  \
topic name                                                                  
python                 9538.0       7162.0       131.0           0.675591   
machine-learning       2239.0        661.0        10.0           0.158592   
hacktoberfest          2098.0       2365.0        34.0           0.148605   
pandas                 1733.0        297.0         5.0           0.122751   
numpy                  1542.0        210.0         3.0           0.109222   
pytorch                1378.0        239.0         9.0           0.097606   
ray                    1146.0          0.0         0.0           0.081173   
data-science            986.0        545.0         9.0           0.069840   
deep-learning           950.0        265.0        10.0           0.067290   
tensorflow              947.0         26.0         1.0           0.067077   
parallel                906.0         77.0         2.0           0.064173   
data-analysis           876.0        280.0         5.0           0.062048   
dataframe               864.0         59.0         1.0           0.061198   
scikit-learn            755.0         64.0         2.0           0.053478   
flexible                736.0          0.0         0.0           0.052132   
alignment               709.0          0.0         0.0           0.050220   
tensor                  686.0          0.0         0.0           0.048590   
compiler                672.0          0.0         0.0           0.047599   
rocm                    658.0          0.0         0.0           0.046607   
cuda                    657.0          0.0         0.0           0.046536   
joblib                  635.0          0.0         0.0           0.044978   
xgboost                 635.0          0.0         0.0           0.044978   
statsmodels             635.0          0.0         0.0           0.044978   
lightgbm                635.0          0.0         0.0           0.044978   
llvm                    607.0         20.0         1.0           0.042995   
linter                  566.0          0.0         0.0           0.040091   
python3                 496.0         97.0         3.0           0.035132   
distributed             490.0          9.0         1.0           0.034707   
ai                      443.0        162.0         3.0           0.031378   
typing                  419.0          0.0         0.0           0.029678   

                  valid topic ratio  test topic ratio  
topic name                                             
python                     0.757323          0.813665  
machine-learning           0.069895          0.062112  
hacktoberfest              0.250079          0.211180  
pandas                     0.031405          0.031056  
numpy                      0.022206          0.018634  
pytorch                    0.025272          0.055901  
ray                        0.000000          0.000000  
data-science               0.057629          0.055901  
deep-learning              0.028022          0.062112  
tensorflow                 0.002749          0.006211  
parallel                   0.008142          0.012422  
data-analysis              0.029608          0.031056  
dataframe                  0.006239          0.006211  
scikit-learn               0.006767          0.012422  
flexible                   0.000000          0.000000  
alignment                  0.000000          0.000000  
tensor                     0.000000          0.000000  
compiler                   0.000000          0.000000  
rocm                       0.000000          0.000000  
cuda                       0.000000          0.000000  
joblib                     0.000000          0.000000  
xgboost                    0.000000          0.000000  
statsmodels                0.000000          0.000000  
lightgbm                   0.000000          0.000000  
llvm                       0.002115          0.006211  
linter                     0.000000      

Во всех выборках хорошо представлен топик python, в train лучше представлен топик machine-learning.

Теперь отсортируем по столбцу `valid topic ratio`.

In [74]:
topics_count.sort_values(by='valid topic ratio',
                         ascending=False).head(30)

train count  valid count  test count  \
topic name                                                        
python                          9538.0       7162.0       131.0   
hacktoberfest                   2098.0       2365.0        34.0   
cloud                            145.0       1960.0         9.0   
infrastructure-as-a-code           0.0       1801.0         7.0   
infrastructure-as-code            12.0       1801.0         7.0   
cloud-management                  11.0       1801.0         7.0   
cloud-provisioning                 0.0       1801.0         7.0   
event-management                   0.0       1801.0         7.0   
infrastructure-automation          0.0       1801.0         7.0   
cloud-providers                    0.0       1801.0         7.0   
zeromq                             0.0       1801.0         7.0   
iot                               36.0       1801.0         7.0   
event-stream                       0.0       1801.0         7.0   
edge                               0.0       1801.0         7.0   
infrastructure-management          0.0       1801.0         7.0   
infrasructure                      0.0       1801.0         7.0   
configuration-management          31.0       1801.0         7.0   
remote-execution                   0.0       1801.0         7.0   
ansible                           63.0       1415.0        15.0   
bittorrent                        56.0        717.0         4.0   
torrent-client                     0.0        717.0         4.0   
privacy                          210.0        717.0         4.0   
search                            39.0        717.0         4.0   
decentralized                     56.0        717.0         4.0   
streaming                         43.0        717.0         4.0   
tor                               39.0        717.0         4.0   
p2p                              106.0        717.0         4.0   
machine-learning                2239.0        661.0        10.0   
bot                               68.0        597.0         4.0   
chatbot                           60.0        569.0         3.0   

                           train topic ratio  valid topic ratio  \
topic name                                                        
python                              0.675591           0.757323   
hacktoberfest                       0.148605           0.250079   
cloud                               0.010271           0.207254   
infrastructure-as-a-code            0.000000           0.190441   
infrastructure-as-code              0.000850           0.190441   
cloud-management                    0.000779           0.190441   
cloud-provisioning                  0.000000           0.190441   
event-management                    0.000000           0.190441   
infrastructure-automation           0.000000           0.190441   
cloud-providers                     0.000000           0.190441   
zeromq                              0.000000           0.190441   
iot                                 0.002550           0.190441   
event-stream                        0.000000           0.190441   
edge                                0.000000           0.190441   
infrastructure-management           0.000000           0.190441   
infrasructure                       0.000000           0.190441   
configuration-management            0.002196           0.190441   
remote-execution                    0.000000           0.190441   
ansible                             0.004462           0.149625   
bittorrent                          0.003967           0.075817   
torrent-client                      0.000000           0.075817   
privacy                             0.014875           0.075817   
search                              0.002762           0.075817   
decentralized                       0.003967           0.075817   
streaming                           0.003046           0.075817   
tor                                 0.002762           0.075817   
p2p   

На валидационной выборке лучше представлены топики 
hacktoberfest

cloud

infrastructure-as-a-code

infrastructure-as-code

cloud-management

cloud-provisioning

event-management

infrastructure-automation

cloud-providers

zeromq

iot

event-stream

edge

infrastructure-management

infrasructure

configuration-management

remote-execution

ansible

Теперь отсортируем по столбцу `test topic ratio`

In [77]:
topics_count.sort_values(by='test topic ratio',
                         ascending=False).head(30)

train count  valid count  test count  train topic ratio  \
topic name                                                                   
python                  9538.0       7162.0       131.0           0.675591   
hacktoberfest           2098.0       2365.0        34.0           0.148605   
automation               239.0        203.0        21.0           0.016929   
threat                     0.0         83.0        17.0           0.000000   
phishing                   0.0         83.0        17.0           0.000000   
incident-response          0.0         83.0        17.0           0.000000   
incident                   0.0         83.0        17.0           0.000000   
handling                   0.0         83.0        17.0           0.000000   
alerts                     3.0         83.0        17.0           0.000212   
ioc                        2.0         83.0        17.0           0.000142   
malware                   20.0         83.0        17.0           0.001417   
ihap                       0.0         83.0        17.0           0.000000   
feeds                      0.0         83.0        17.0           0.000000   
cybersecurity              0.0         83.0        17.0           0.000000   
intelligence               0.0         83.0        17.0           0.000000   
csirt                      0.0         83.0        17.0           0.000000   
cert                       0.0         83.0        17.0           0.000000   
ansible                   63.0       1415.0        15.0           0.004462   
discord                   52.0        465.0        11.0           0.003683   
computer-vision          149.0         21.0        10.0           0.010554   
image-processing          29.0         21.0        10.0           0.002054   
deep-learning            950.0        265.0        10.0           0.067290   
machine-learning        2239.0        661.0        10.0           0.158592   
python-3                  53.0         25.0         9.0           0.003754   
cloud                    145.0       1960.0         9.0           0.010271   
pytorch                 1378.0        239.0         9.0           0.097606   
data-science             986.0        545.0         9.0           0.069840   
discord-api                5.0         25.0         9.0           0.000354   
support                    2.0         38.0         8.0           0.000142   
erp                        1.0         38.0         8.0           0.000071   

                   valid topic ratio  test topic ratio  
topic name                                              
python                      0.757323          0.813665  
hacktoberfest               0.250079          0.211180  
automation                  0.021466          0.130435  
threat                      0.008777          0.105590  
phishing                    0.008777          0.105590  
incident-response           0.008777          0.105590  
incident                    0.008777          0.105590  
handling                    0.008777          0.105590  
alerts                      0.008777          0.105590  
ioc                         0.008777          0.105590  
malware                     0.008777          0.105590  
ihap                        0.008777          0.105590  
feeds                       0.008777          0.105590  
cybersecurity               0.008777          0.105590  
intelligence                0.008777          0.105590  
csirt                       0.008777          0.105590  
cert                        0.008777          0.105590  
ansible                     0.149625          0.093168  
discord                     0.049170          0.068323  
computer-vision             0.002221          0.062112  
image-processing            0.002221          0.062112  
deep-learning               0.028022          0.062112  
machine-learning            0.069895          0.062112  
python-3                    0.002644          0.055901  
cloud                       0.207254    

В тестовой выборке хорошо представлены тематики:

automation

threat

phishing

incident-response

incident

handling

alerts

ioc

malware

ihap

feeds

cybersecurity

intelligence

csirt

cert

ansible

# Загрузка данных по корректному коду

Загрузим данные всех датасетов

In [124]:
source_code_data_train_ = pd.read_pickle("correct_source_code_train.pickle")
source_code_data_valid_ = pd.read_pickle("correct_source_code_valid.pickle")
source_code_data_test_ = pd.read_pickle("correct_source_code_test.pickle")

In [125]:
train_repos_ = load_repo_names(source_code_data_train_)
valid_repos_ = load_repo_names(source_code_data_valid_)
test_repos_ = load_repo_names(source_code_data_test_)

Посмотрим сколько репозиториев представлено в корректном коде.

In [126]:
len(test_repos_), len(train_repos_), len(valid_repos_), len(test_repos_) + len(train_repos_) + len(valid_repos_)

(94, 1571, 10536, 12201)

Посмотрим сколько сниппетов лежит в каждом репозитории

In [127]:
train_repos_count_ = source_code_data_train_['repo_name'].value_counts()

for i, count in zip(train_repos_count_.index, train_repos_count_):
    print(i, count)

cloudera_hue 13660
aliyun_aliyun-openapi-python-sdk 10261
Azure_azure-cli-extensions 9139
dimagi_commcare-hq 7532
XX-net_XX-Net 6384
AppScale_gts 5782
openhatch_oh-mainline 4976
cryoem_eman2 4948
morpheus65535_bazarr 4819
napalm-automation_napalm-yang 4662
Jenyay_outwiker 3444
spack_spack 3298
vivisect_vivisect 3175
openstack_neutron 2697
ms-iot_python 2607
pandas-dev_pandas 2588
tensorflow_models 2403
nikssardana_gdriveTorrent 2370
indico_indico 2316
learningequality_ka-lite 2294
Nuitka_Nuitka 2275
theotherp_nzbhydra 2180
dagster-io_dagster 2092
faucetsdn_ryu 1984
Arelle_Arelle 1942
nerdvegas_rez 1852
makerbot_ReplicatorG 1831
superdesk_Live-Blog 1796
emesene_emesene 1796
spesmilo_electrum 1758
Arx-Game_arxcode 1594
smarthomeNG_plugins 1548
Adapter-Hub_adapter-transformers 1546
sqlalchemy_sqlalchemy 1464
pyrocko_pyrocko 1445
keiffster_program-y 1412
enthought_pyface 1392
gentoo_portage 1379
gramps-project_addons-source 1376
oilshell_oil 1364
PaddlePaddle_models 1359
OpenNetworkingFoun

dask_dask-glm 42
Devoxin_Lavalink.py 42
MITgcm_xmitgcm 42
mswart_openmensa-parsers 42
miguelgrinberg_python-engineio 42
GenericMappingTools_pygmt 42
friggog_tree-gen 42
codeshow_curso-flask 42
gawel_panoramisk 42
frankie567_fastapi-users 42
23andMe_Yamale 41
megagonlabs_ginza 41
HASecuritySolutions_VulnWhisperer 41
Ganapati_RsaCtfTool 41
dalibo_pg_activity 41
nighres_nighres 41
fwestenberg_reolink_dev 41
mosquito_aiormq 41
ishepard_pydriller 41
httplib2_httplib2 41
gawel_pyquery 41
rbarrois_python-semanticversion 41
model-bakers_model_bakery 41
zhanghang1989_ResNeSt 41
OpenLMIS_openlmis-ref-distro 40
witten_borgmatic 40
SmileyChris_django-countries 40
alfredfrancis_ai-chatbot-framework 40
quantopian_empyrical 40
SmileyChris_easy-thumbnails 40
ztgrace_changeme 40
Tivix_django-common 39
EricJMarti_inventory-hunter 39
pmorissette_ffn 39
fieldOfView_Cura-OctoPrintPlugin 39
lisa-lab_DeepLearningTutorials 39
odoo-brazil_odoo-brazil-eletronic-documents 39
PathagarBooks_pathagar 39
frida_frida

In [128]:
valid_repos_count_ = source_code_data_valid_['repo_name'].value_counts()

for i, count in zip(valid_repos_count_.index, valid_repos_count_):
    print(i, count)

CiscoDevNet_ydk-py 254270
Azure_azure-sdk-for-python 88170
oracle_oci-python-sdk 42232
sagemath_sage 40024
tytusdb_tytus 35469
home-assistant_core 28861
cctbx_cctbx_project 26873
tencentyun_scf-demo-repo 25953
cloudera_hue 25299
sympy_sympy 25241
kovidgoyal_calibre 20383
docusign_docusign-python-client 19448
saltstack_salt 18958
aliyun_aliyun-openapi-python-sdk 18938
leo-editor_leo-editor 18735
tribe29_checkmk 18454
Azure_azure-cli-extensions 16774
Toontown-Open-Source-Initiative_Toontown-School-House 16696
anhstudios_swganh 15709
SteveDoyle2_pyNastran 15240
dnanexus_parliament2 14771
plotly_plotly.py 14043
wxWidgets_Phoenix 13965
dimagi_commcare-hq 13886
demisto_content 13746
SickChill_SickChill 13669
galaxyproject_galaxy 12703
gramps-project_gramps 11575
XX-net_XX-Net 11521
openstack_cinder 11201
AppScale_gts 10369
VOLTTRON_volttron 9972
ActiveState_code 9946
edx_edx-platform 9817
tomplus_kubernetes_asyncio 9434
nortikin_sverchok 9381
openhatch_oh-mainline 9116
morpheus65535_bazarr 9

FreeOpcUa_python-opcua 1148
aws_chalice 1147
CiscoUcs_imcsdk 1146
wolverton-research-group_qmpy 1146
scrapy_scrapy 1144
google_corpuscrawler 1143
opencybersecurityalliance_stix-shifter 1141
facebookresearch_nevergrad 1140
brad-sp_cuckoo-modified 1139
Project-MONAI_MONAI 1138
RaRe-Technologies_gensim 1138
rte-france_Grid2Op 1137
liqd_adhocracy 1137
openfisca_openfisca-france 1136
ansible-collections_community.vmware 1135
controlscanmdr_cyphon 1135
asyml_texar 1135
rockstor_rockstor-core 1135
ricequant_rqalpha 1131
USEPA_WNTR 1129
Neuraxio_Neuraxle 1128
robotools_fontParts 1127
pyvista_pyvista 1125
rubys_feedvalidator 1125
Scifabric_pybossa 1124
mementum_backtrader 1124
osm-fr_osmose-backend 1123
anki_cozmo-python-sdk 1121
online-ml_river 1121
wkeeling_selenium-wire 1120
aiogram_aiogram 1120
beetbox_beets 1119
fofix_fofix 1118
multiSnow_mcomix3 1117
skelsec_pypykatz 1115
Terbau_fortnitepy 1114
diefenbach_django-lfs 1113
pmgbergen_porepy 1113
westpa_westpa 1113
tensorflow_model-analysis 1

spacetx_starfish 685
googlefonts_glyphsLib 685
gramener_gramex 684
facebookresearch_pyrobot 684
IDSIA_brainstorm 684
WZMIAOMIAO_deep-learning-for-image-processing 683
fireeye_capa 683
emc-isilon_pike 683
MediaBrowser_plugin.video.emby 683
bup_bup 682
PyCQA_astroid 682
jendrikseipp_rednotebook 682
PaddlePaddle_benchmark 680
m-labs_migen 679
pyreadline_pyreadline 679
praw-dev_praw 679
QData_TextAttack 679
usnistgov_jarvis 678
pypest_pyemu 678
vit-project_vit 678
conjure-up_conjure-up 678
intel_workload-collocation-agent 677
dephell_dephell 677
jansel_opentuner 676
openrazer_openrazer 675
WPO-Foundation_wptagent 675
BigBrotherBot_big-brother-bot 675
oemof_tespy 675
IndicoDataSolutions_finetune 675
mila-iqia_blocks 675
reddit_baseplate.py 674
HewlettPackard_oneview-ansible 674
willmcgugan_rich 674
gaasedelen_lighthouse 673
Netflix_vmaf 672
intel_dffml 672
fooying_3102 672
openai_jukebox 672
joelgrus_data-science-from-scratch 672
inducer_pyopencl 671
hakyimlab_MetaXcan 671
pypa_virtualenv 6

NVIDIA_apex 437
nabaztag2018_pynab 437
ceph_ceph-qa-suite 436
mikgroup_sigpy 436
lutris_website 436
Tanganelli_CoAPthon 436
aliyun_aliyun-oss-python-sdk 436
ansible-collections_ibm_zos_core 436
databricks_databricks-cli 435
GeoscienceAustralia_anuga_core 435
Ulauncher_Ulauncher 435
vim-vdebug_vdebug 435
Flagsmith_flagsmith-api 435
theislab_scvelo 434
tensorflow_data-validation 434
tomerfiliba-org_rpyc 434
trigger_trigger 434
psf_pyperf 433
kvesteri_sqlalchemy-utils 432
user-cont_conu 432
VIDA-NYU_reprozip 432
creativecommons_cccatalog 432
sanderland_katrain 432
InstituteforDiseaseModeling_covasim 432
dataloop-ai_AutoML 432
regro_cf-scripts 432
eNMS-automation_eNMS 432
aiortc_aioquic 431
antoinecarme_pyaf 431
hamish2014_FreeCAD_assembly2 431
TitanEmbeds_Titan 431
HDFGroup_hdf-compass 430
awslabs_aws-config-rules 430
Cullyege_Hacktoberfest2020 430
angr_pyvex 430
SoftwareQuTech_SimulaQron 429
tensorflow_ranking 429
honeynet_droidbot 429
deepmind_pysc2 429
albermax_innvestigate 429
m-labs_

pytransitions_transitions 285
Lab41_Circulo 285
Mirantis_kqueen 285
jimmysong_programmingbitcoin 285
sequitur-g2p_sequitur-g2p 285
QingdaoU_OnlineJudge 285
ihabunek_toot 285
Finance-Hub_FinanceHub 285
OCA_manufacture 285
paris-saclay-cds_ramp-workflow 284
johannfaouzi_pyts 284
tdryer_hangups 284
mlpack_benchmarks 284
ActiDoo_gamification-engine 284
MultiAgentLearning_playground 284
authomatic_authomatic 284
Noctem_Monocle 284
python-gitlab_python-gitlab 284
steven-murray_hmf 284
LLNL_maestrowf 283
jamesmeneghello_pynab 283
cubicle-vdo_xbmc-israel 283
snowflakedb_SnowAlert 283
Vaelor_python-mattermost-driver 282
christiansandberg_canopen 282
sripathikrishnan_redis-rdb-tools 282
nutti_fake-bpy-module 282
blackjack4494_yt-dlc 282
google_vulncode-db 282
jaegertracing_jaeger-client-python 282
robinhood_thorn 282
choderalab_yank 282
turicas_rows 282
openlawlibrary_pygls 282
fbchat-dev_fbchat 281
e3nn_e3nn 281
aws-cloudformation_cloudformation-cli 281
grin-pool_grin-pool 281
dedupeio_dedupe 2

fabric-bolt_fabric-bolt 183
sepandhaghighi_pycm 183
EnableSecurity_wafw00f 183
boppreh_keyboard 183
dropbox_pynsot 183
openscopeproject_InteractiveHtmlBom 182
Bluefissure_OtterBot 182
ezaquarii_vpn-at-home 182
cagbal_ros_people_object_detection_tensorflow 182
andriyko_sublime-robot-framework-assistant 182
zopefoundation_z3c.rml 182
guessit-io_guessit 182
mozilla_source 182
ultralytics_yolov3 182
0rpc_zerorpc-python 182
googlefonts_ufo2ft 182
kivy_kivy-ios 182
djbook-ru_django_documentation 182
r9y9_deepvoice3_pytorch 182
funtoo_merge-scripts 182
gmr_consulate 181
CharlesShang_FastMaskRCNN 181
geometalab_OSMDeepOD 181
nasa_RHEAS 181
adrienverge_localstripe 181
openlabs_nereid 181
jazzband-roadies_website 181
serhatbolsu_robotframework-appiumlibrary 181
blabla1337_skf-labs 181
ms-iot_iot-utilities 181
readthedocs_sphinx-autoapi 181
stan-dev_cmdstanpy 181
modoboa_modoboa-webmail 181
quantopian_pgcontents 181
JelteF_PyLaTeX 181
KenT2_tboplayer 181
paypal_PPExtensions 181
algoo_preview-gene

cihologramas_pyoptools 125
benoitc_restkit 125
weecology_DeepForest 125
elastic_ecs 125
prometheus_client_python 125
WilldooIT_Pentaho-reports-for-OpenERP 125
olofk_edalize 125
dell_OpenManage-Enterprise 125
camelot-dev_camelot 125
kieferk_dfply 125
aewallin_allantools 125
fcurella_django-recommends 125
ehmatthes_pcc 125
julython_julython.org 125
imWildCat_scylla 125
locationlabs_mockredis 125
neurodsp-tools_neurodsp 125
flapjax_FlapJack-Cogs 125
dtcooper_python-fitparse 125
zeshuaro_telegram-pdf-bot 125
Nextdoor_ndscheduler 125
adfinis-sygroup_foreman-yml 125
dmitry-k_yandex_smart_home 125
LettError_MutatorMath 125
scikit-learn-contrib_imbalanced-learn 125
trakt_script.trakt 125
nryoung_algorithms 125
bepasty_bepasty-server 125
openworm_owmeta 124
jrfonseca_xdot.py 124
cf-platform-eng_tile-generator 124
viper-framework_viper 124
Ishaan28malik_Hacktoberfest-2020 124
disqus_nydus 124
openparty_openparty 124
ajdawson_eofs 124
materialsvirtuallab_pymatgen-diffusion 124
saghul_pythonz 124


UDST_urbanaccess 85
mishbahr_djangocms-forms 85
dusktreader_flask-praetorian 85
datawire_forge 85
UCBerkeleySETI_blimpy 85
APSL_django-yubin 85
kurtbrose_pyjks 85
tarmstrong_nbdiff 85
ceph_cephmetrics 85
alpacahq_alpaca-backtrader-api 85
django-dbbackup_django-dbbackup 85
sdelements_django-security 85
aws_deep-learning-containers 85
alexa_alexa-smarthome 85
humberry_ui-tutorial 85
getsentry_rb 85
jfkirk_tensorrec 85
explosion_sense2vec 85
glm-tools_pyglmnet 84
faircloth-lab_phyluce 84
rpm-software-management_dnf-plugins-extras 84
C-Otto_rebalance-lnd 84
callowayproject_django-categories 84
spotify_snakebite 84
konlpy_konlpy 84
infrae_pyoai 84
PyCQA_pylint-django 84
TheRenegadeCoder_sample-programs 84
allenai_scispacy 84
fusionbox_django-betterforms 84
KordingLab_spykes 84
ros2_examples 84
evildmp_BrachioGraph 84
amundsen-io_amundsensearchlibrary 84
amluto_virtme 84
mailgun_talon 84
panzarino_mlbgame 84
astropy_ccdproc 84
jendrikseipp_vulture 84
dolphin-emu_www 84
JBKahn_django-sharding

berinhard_model_mommy 52
openai_universe-starter-agent 51
jmcarpenter2_swifter 51
Mergifyio_git-pull-request 51
danhenriquesc_hackerrank 51
bambinos_bambi 51
appnexus_pyrobuf 51
divio_cmsplugin-filer 51
applied-bioinformatics_An-Introduction-To-Applied-Bioinformatics 51
eltonlaw_impyute 51
segmentio_analytics-python 51
voxpupuli_puppetboard 51
timvideos_streaming-system 51
intjelic_python-sfml 51
chakki-works_seqeval 51
jarondl_pygtfs 51
sass_libsass-python 51
FeeiCN_ESD 51
pydata_pandas-gbq 51
donnemartin_viz 51
Hyperclaw79_PokeBall-SelfBot-Pokecord-Automation 51
singer-io_tap-hubspot 51
ymattw_ydiff 51
pypxe_PyPXE 51
lorenzodifuccia_safaribooks 51
korfuri_django-prometheus 51
spyder-ide_qtawesome 51
kislyuk_argcomplete 51
openai_gym-http-api 51
zulip_zulip-gci 51
akalongman_sublimetext-stringutilities 51
pypa_wheel 51
pydanny_cookiecutter-django 51
rm-hull_luma.led_matrix 51
imcaspar_gpt2-ml 51
ethereum_beacon_chain 51
dbader_schedule 51
decalage2_olefile 51
napalm-automation_napalm-

unbit_uwsgi-docs 28
joke2k_django-environ 28
swaroopch_byte-of-python 28
mayukh18_BlindChat 28
django-ses_django-ses 28
bit101_ProjectMaker 28
XML-Security_signxml 28
vyos_vyos-documentation 28
miracle2k_tarsnapper 28
cyprieng_swagger-parser 28
nkgilley_python-ecobee-api 28
alex_letsencrypt-aws 28
newsapps_beeswithmachineguns 28
atlanhq_camelot 28
Tinche_aiofiles 28
thiderman_doge 28
torchbox_wagtail-markdown 28
mackorone_spotify-playlist-archive 28
MissionCriticalCloud_cloudstackOps 28
rackerlabs_openstack-guest-agents-unix 28
kholia_OSX-KVM 28
maximdanilchenko_aiohttp-apispec 28
hzdg_django-enumfields 28
SJTU-CSE_awesome-cs 28
lyft_awspricing 28
Tinche_cattrs 28
brennerm_PyTricks 28
invoice-x_invoice2data 28
fabpot_sphinx-php 28
pnuckowski_aioresponses 28
TwilioDevEd_account-security-quickstart-django 28
pycco-docs_pycco 28
biocore_deblur 28
YunYang1994_tensorflow-yolov3 28
phalt_swapi 28
CamDavidsonPilon_tdigest 28
encode_django-vanilla-views 28
OrganicIrradiation_ratebeer 28
armooo

minimaxir_facebook-page-post-scraper 10
Anupam-dagar_Portfolio-Generator 10
mubaris_motivate 10
pytorch_hub 10
agronholm_sphinx-autodoc-typehints 10
byte-stream_ARTS 10
DannyCork_python-whois 10
ZeevG_python-forecast.io 10
Turbo87_utm 10
lepture_mistune 10
JakeWharton_pidcat 10
openzfs_openzfs-docs 10
dcramer_django-uuidfield 10
stefanofontanelli_ColanderAlchemy 10
Leviathan1995_Pylsy 10
hellysmile_django-activeurl 10
wemake-services_wemake-django-template 10
azavea_django-queryset-csv 10
bhrigu123_classifier 10
lionheart_bottlenose 10
avinassh_rockstar 10
cosmicpython_book 10
praekeltfoundation_django-google-analytics 10
torproject_torspec 10
aparrish_pronouncingpy 10
redhat-cop_openshift-management 10
studio-ousia_mprpc 10
mjhea0_flaskr-tdd 10
Diblo_KODI-Popcorn-Time 10
verigak_progress 10
rougier_numpy-100 10
jadchaar_sec-edgar-downloader 10
fivesheep_chnroutes 10
datadriventests_ddt 10
shauryauppal_PyWhatsapp 10
bitcoinbook_bitcoinbook 10
qiwsir_algorithm 10
jupyterhub_sudospawner 

In [129]:
test_repos_count_ = source_code_data_test_['repo_name'].value_counts()

for i, count in zip(test_repos_count_.index, test_repos_count_):
    print(i, count)

graalvm_mx 12
andikleen_pmu-tools 9
googlefonts_noto-emoji 6
alexbers_mtprotoproxy 6
emscripten-core_emsdk 5
hugsy_gef 5
Bob123a1_CDNSP-GUI 4
ClangBuiltLinux_tc-build 4
animate1978_MB-Lab 4
belguawhale_Discord-Werewolf 3
DestructHub_ProjectEuler 3
Charcoal-SE_SmokeDetector 3
hhatto_autopep8 3
GlacierProtocol_GlacierProtocol 3
21isenough_LightningATM 3
elyra-ai_elyra 2
edwardgeorge_virtualenv-clone 2
Felienne_hedy 2
darkoperator_dnsrecon 2
autocracy_python-ipy 2
ekalinin_nodeenv 2
ArduPilot_pymavlink 2
aziz_PlainTasks 2
georgewhewell_undervolt 2
google-research_bert 2
frej_fast-export 2
IgnoredAmbience_yahoo-group-archiver 2
aio-libs_aioftp 2
facebookarchive_python-instagram 2
gregzaal_Auto-Voice-Channels 2
DataDog_dd-agent 2
containers_podman-compose 2
GoogleChrome_chromium-dashboard 2
aachman98_Sorcar 1
erpalma_throttled 1
google-research_albert 1
fr31_spotifylyrics 1
gooofy_zamia-speech 1
gornostal_Modific 1
gangverk_flask-swagger 1
e-dard_flask-s3 1
0x5e_wechat-deleted-friends 1
goo

Сниппеты обучающей и тестовой выборок не пересекаются по репозиториям:

In [130]:
set(test_repos_) & set(train_repos_)

set()

Репозитории в валидационной и тестовой выборках также не пересекаются:

In [131]:
set(test_repos_) & set(valid_repos_)

set()

# Топики репозиториев

Загрузим данные с топиками репозиториев для кода с ошибками.

In [132]:
repo_name_to_topics = pickle.load(open("topics.pickle", "rb"))
train_repos_topics_ = [repo_name_to_topics[repo_name] for repo_name in source_code_data_train_['repo_name'].to_numpy() if repo_name in repo_name_to_topics]
valid_repos_topics_ = [repo_name_to_topics[repo_name] for repo_name in source_code_data_valid_['repo_name'].to_numpy()  if repo_name in repo_name_to_topics]
test_repos_topics_ = [repo_name_to_topics[repo_name] for repo_name in source_code_data_test_['repo_name'].to_numpy()  if repo_name in repo_name_to_topics]

In [133]:
train_repos_topics_cnt_ = collections.Counter()
for t in train_repos_topics_:
    train_repos_topics_cnt_.update(set(t))
    
valid_repos_topics_cnt_ = collections.Counter()
for t in valid_repos_topics_:
    valid_repos_topics_cnt_.update(set(t))

test_repos_topics_cnt_ = collections.Counter()
for t in test_repos_topics_:
    test_repos_topics_cnt_.update(set(t))

Наиболее представленные топики в обучающей выборке:

In [134]:
train_repos_topics_cnt_.most_common()

[('python', 106168),
 ('hacktoberfest', 16137),
 ('sql', 15590),
 ('machine-learning', 13924),
 ('autocomplete', 13691),
 ('data-warehouse', 13660),
 ('query-editor', 13660),
 ('compose', 13660),
 ('databases', 13660),
 ('sql-editor', 13660),
 ('sql-assistant', 13660),
 ('deep-learning', 12863),
 ('sdk', 10664),
 ('alibaba-cloud-sdk', 10261),
 ('alibaba-cloud-python', 10261),
 ('development-kit', 10261),
 ('python3', 10015),
 ('azure-cli-extension', 9139),
 ('azure-cli', 9139),
 ('linux', 8307),
 ('pytorch', 7828),
 ('commcare-hq', 7532),
 ('django', 7194),
 ('data-science', 6958),
 ('tensorflow', 6864),
 ('proxy', 6509),
 ('vpn', 6436),
 ('gfw', 6384),
 ('uncensored', 6384),
 ('goagent', 6384),
 ('flask', 6034),
 ('security', 4886),
 ('reverse-engineering', 4854),
 ('radarr', 4836),
 ('sonarr', 4836),
 ('nlp', 4822),
 ('series', 4819),
 ('episodes', 4819),
 ('subtitle-downloader', 4819),
 ('download-subtitles', 4819),
 ('subliminal', 4819),
 ('movie', 4819),
 ('natural-language-proces

Наиболее представленные топики в валидационной выборке:

In [135]:
valid_repos_topics_cnt_.most_common()

[('python', 2138474),
 ('hacktoberfest', 388057),
 ('machine-learning', 273116),
 ('catalyst', 254982),
 ('yang', 254888),
 ('ydk', 254674),
 ('ios-xe', 254327),
 ('ios-xr', 254309),
 ('nx-os', 254309),
 ('deep-learning', 205297),
 ('python3', 145391),
 ('science', 140602),
 ('pytorch', 136175),
 ('django', 130053),
 ('tensorflow', 118520),
 ('azure', 114756),
 ('linux', 102777),
 ('data-science', 101798),
 ('cloud', 98132),
 ('microsoft', 94441),
 ('azure-sdk', 88302),
 ('sdk', 86825),
 ('security', 78570),
 ('asyncio', 78444),
 ('visualization', 78226),
 ('docker', 71510),
 ('computer-algebra', 67886),
 ('javascript', 67562),
 ('nlp', 66523),
 ('math', 66430),
 ('numpy', 64166),
 ('iot', 61351),
 ('reinforcement-learning', 57042),
 ('natural-language-processing', 54989),
 ('workflow', 54629),
 ('automation', 52517),
 ('raspberry-pi', 51650),
 ('windows', 51349),
 ('pandas', 48091),
 ('aws', 47160),
 ('algorithms', 45731),
 ('home-automation', 44510),
 ('mathematics', 44114),
 ('neura

Наиболее представленные топики в тестовой выборке:

In [136]:
test_repos_topics_cnt_.most_common()

[('python', 51),
 ('docker', 10),
 ('linux', 8),
 ('hacktoberfest', 8),
 ('proxy', 6),
 ('mtproto', 6),
 ('async-mtproto-proxy', 6),
 ('telegram', 6),
 ('mtproxy', 6),
 ('gdb', 5),
 ('ctf', 5),
 ('exploit', 5),
 ('binary-ninja', 5),
 ('reverse-engineering', 5),
 ('pwn', 5),
 ('ida-pro', 5),
 ('exploit-development', 5),
 ('debugging', 5),
 ('malware-analysis', 5),
 ('blender-addon', 5),
 ('blender', 5),
 ('addon', 5),
 ('machine-learning', 5),
 ('manuelbastionilab', 4),
 ('discord', 3),
 ('mafia', 3),
 ('game', 3),
 ('bot', 3),
 ('werewolf', 3),
 ('samples', 3),
 ('undervolt', 3),
 ('tensorflow', 3),
 ('metrics', 3),
 ('web', 3),
 ('regex', 3),
 ('spam', 3),
 ('ai', 3),
 ('projecteuler', 3),
 ('haskell', 3),
 ('common-lisp', 3),
 ('contests', 3),
 ('hackertoberfest', 3),
 ('solvers', 3),
 ('polyglot', 3),
 ('lua', 3),
 ('mathematics', 3),
 ('cplusplus-11', 3),
 ('elixir', 3),
 ('solver', 3),
 ('clojure', 3),
 ('devops', 3),
 ('pep8', 3),
 ('formatter', 3),
 ('codeformatter', 3),
 ('subl

Сравним представленности топиков:

In [137]:
train_topics_count_ = pd.DataFrame(train_repos_topics_cnt_.most_common())
train_topics_count_.rename({0: 'topic name', 1: 'train count'},
                           inplace=True,
                           axis=1)
train_topics_count_.set_index(keys='topic name', inplace=True)

valid_topics_count_ = pd.DataFrame(valid_repos_topics_cnt_.most_common())
valid_topics_count_.rename({0: 'topic name', 1: 'valid count'},
                           inplace=True,
                           axis=1)
valid_topics_count_.set_index(keys='topic name', 
                              inplace=True)
test_topics_count_ = pd.DataFrame(test_repos_topics_cnt_.most_common())
test_topics_count_.rename({0: 'topic name', 1: 'test count'},
                           inplace=True,
                           axis=1)
test_topics_count_.set_index(keys='topic name', inplace=True)

In [138]:
topics_count_ = pd.concat((train_topics_count_, valid_topics_count_, test_topics_count_), 
                           axis=1)

In [139]:
topics_count_['train topic ratio'] = topics_count_['train count'] / source_code_data_train_.shape[0]
topics_count_['valid topic ratio'] = topics_count_['valid count'] / source_code_data_valid_.shape[0]
topics_count_['test topic ratio'] = topics_count_['test count'] / source_code_data_test_.shape[0]

In [140]:
topics_count_ = topics_count_.fillna(0)
topics_count_

train count  valid count  test count  train topic ratio  \
topic name                                                                     
python                  106168.0    2138474.0        51.0           0.301689   
hacktoberfest            16137.0     388057.0         8.0           0.045855   
sql                      15590.0      37050.0         0.0           0.044301   
machine-learning         13924.0     273116.0         5.0           0.039567   
autocomplete             13691.0      25752.0         0.0           0.038905   
...                          ...          ...         ...                ...   
programmable-engine          0.0          0.0         1.0           0.000000   
speech-corpora               0.0          0.0         1.0           0.000000   
cmu-sphinx                   0.0          0.0         1.0           0.000000   
sequitur                     0.0          0.0         1.0           0.000000   
voxforge                     0.0          0.0         1.0           0.000000   

                     valid topic ratio  test topic ratio  
topic name                                                
python                        0.394271          0.300000  
hacktoberfest                 0.071546          0.047059  
sql                           0.006831          0.000000  
machine-learning              0.050354          0.029412  
autocomplete                  0.004748          0.000000  
...                                ...               ...  
programmable-engine           0.000000          0.005882  
speech-corpora                0.000000          0.005882  
cmu-sphinx                    0.000000          0.005882  
sequitur                      0.000000          0.005882  
voxforge                      0.000000          0.005882  

[14165 rows x 6 columns]

Отсортируем по столбцу `train topic ratio`.

In [141]:
topics_count_.sort_values(by='train topic ratio',
                          ascending=False).head(30)

train count  valid count  test count  train topic ratio  \
topic name                                                                      
python                   106168.0    2138474.0        51.0           0.301689   
hacktoberfest             16137.0     388057.0         8.0           0.045855   
sql                       15590.0      37050.0         0.0           0.044301   
machine-learning          13924.0     273116.0         5.0           0.039567   
autocomplete              13691.0      25752.0         0.0           0.038905   
data-warehouse            13660.0      25960.0         0.0           0.038817   
query-editor              13660.0      25299.0         0.0           0.038817   
compose                   13660.0      25299.0         0.0           0.038817   
databases                 13660.0      26284.0         0.0           0.038817   
sql-editor                13660.0      25299.0         0.0           0.038817   
sql-assistant             13660.0      25299.0         0.0           0.038817   
deep-learning             12863.0     205297.0         1.0           0.036552   
sdk                       10664.0      86825.0         0.0           0.030303   
development-kit           10261.0      18938.0         0.0           0.029158   
alibaba-cloud-sdk         10261.0      18938.0         0.0           0.029158   
alibaba-cloud-python      10261.0      18938.0         0.0           0.029158   
python3                   10015.0     145391.0         2.0           0.028459   
azure-cli-extension        9139.0      18493.0         0.0           0.025970   
azure-cli                  9139.0      26304.0         0.0           0.025970   
linux                      8307.0     102777.0         8.0           0.023605   
pytorch                    7828.0     136175.0         0.0           0.022244   
commcare-hq                7532.0      13886.0         0.0           0.021403   
django                     7194.0     130053.0         0.0           0.020443   
data-science               6958.0     101798.0         0.0           0.019772   
tensorflow                 6864.0     118520.0         3.0           0.019505   
proxy                      6509.0      17303.0         6.0           0.018496   
vpn                        6436.0      13779.0         0.0           0.018289   
gfw                        6384.0      11521.0         0.0           0.018141   
uncensored                 6384.0      11521.0         0.0           0.018141   
goagent                    6384.0      11521.0         0.0           0.018141   

                      valid topic ratio  test topic ratio  
topic name                                                 
python                         0.394271          0.300000  
hacktoberfest                  0.071546          0.047059  
sql                            0.006831          0.000000  
machine-learning               0.050354          0.029412  
autocomplete                   0.004748          0.000000  
data-warehouse                 0.004786          0.000000  
query-editor                   0.004664          0.000000  
compose                        0.004664          0.000000  
databases                      0.004846          0.000000  
sql-editor                     0.004664          0.000000  
sql-assistant                  0.004664          0.000000  
deep-learning                  0.037851          0.005882  
sdk                            0.016008          0.000000  
development-kit                0.003492          0.000000  
alibaba-cloud-sdk              0.003492          0.000000  
alibaba-cloud-python           0.003492          0.000000  
python3                        0.026806          0.011765  
azure-cli-extension            0.003410          0.000000  
azure-cli                      0.004850          0.000000  
linux                          0.018949          0.047059  
pytorch                        0.025107          0.000000  
commcare-hq                    0.002560          

Во всех выборках хорошо представлен топик python, в train лучше представлен топик sql.

Теперь отсортируем по столбцу `valid topic ratio`.

In [142]:
topics_count_.sort_values(by='valid topic ratio',
                          ascending=False).head(30)

train count  valid count  test count  train topic ratio  \
topic name                                                                  
python               106168.0    2138474.0        51.0           0.301689   
hacktoberfest         16137.0     388057.0         8.0           0.045855   
machine-learning      13924.0     273116.0         5.0           0.039567   
catalyst                193.0     254982.0         0.0           0.000548   
yang                      0.0     254888.0         0.0           0.000000   
ydk                       0.0     254674.0         0.0           0.000000   
ios-xe                    0.0     254327.0         0.0           0.000000   
nx-os                     0.0     254309.0         0.0           0.000000   
ios-xr                    0.0     254309.0         0.0           0.000000   
deep-learning         12863.0     205297.0         1.0           0.036552   
python3               10015.0     145391.0         2.0           0.028459   
science                2234.0     140602.0         0.0           0.006348   
pytorch                7828.0     136175.0         0.0           0.022244   
django                 7194.0     130053.0         0.0           0.020443   
tensorflow             6864.0     118520.0         3.0           0.019505   
azure                  1083.0     114756.0         0.0           0.003077   
linux                  8307.0     102777.0         8.0           0.023605   
data-science           6958.0     101798.0         0.0           0.019772   
cloud                   659.0      98132.0         0.0           0.001873   
microsoft               113.0      94441.0         0.0           0.000321   
azure-sdk                 0.0      88302.0         0.0           0.000000   
sdk                   10664.0      86825.0         0.0           0.030303   
security               4886.0      78570.0         1.0           0.013884   
asyncio                3185.0      78444.0         2.0           0.009051   
visualization          1934.0      78226.0         0.0           0.005496   
docker                 3884.0      71510.0        10.0           0.011037   
computer-algebra          0.0      67886.0         0.0           0.000000   
javascript             1802.0      67562.0         1.0           0.005121   
nlp                    4822.0      66523.0         2.0           0.013702   
math                     93.0      66430.0         0.0           0.000264   

                  valid topic ratio  test topic ratio  
topic name                                             
python                     0.394271          0.300000  
hacktoberfest              0.071546          0.047059  
machine-learning           0.050354          0.029412  
catalyst                   0.047011          0.000000  
yang                       0.046994          0.000000  
ydk                        0.046954          0.000000  
ios-xe                     0.046890          0.000000  
nx-os                      0.046887          0.000000  
ios-xr                     0.046887          0.000000  
deep-learning              0.037851          0.005882  
python3                    0.026806          0.011765  
science                    0.025923          0.000000  
pytorch                    0.025107          0.000000  
django                     0.023978          0.000000  
tensorflow                 0.021852          0.017647  
azure                      0.021158          0.000000  
linux                      0.018949          0.047059  
data-science               0.018769          0.000000  
cloud                      0.018093          0.000000  
microsoft                  0.017412          0.000000  
azure-sdk                  0.016280          0.000000  
sdk                        0.016008          0.000000  
security                   0.014486          0.005882  
asyncio                    0.014463          0.011765  
visualization              0.014423          0.000000  
docker                     0.013184      

На валидационной выборке лучше представлены топики 

hacktoberfest

machine-learning

catalyst

yang

ydk

ios-x

nx-os

ios-xr

Теперь отсортируем по столбцу `test topic ratio`

In [143]:
topics_count_.sort_values(by='test topic ratio',
                         ascending=False).head(30)

train count  valid count  test count  train topic ratio  \
topic name                                                                     
python                  106168.0    2138474.0        51.0           0.301689   
docker                    3884.0      71510.0        10.0           0.011037   
linux                     8307.0     102777.0         8.0           0.023605   
hacktoberfest            16137.0     388057.0         8.0           0.045855   
telegram                   822.0      11488.0         6.0           0.002336   
mtproxy                      0.0          0.0         6.0           0.000000   
async-mtproto-proxy          0.0          0.0         6.0           0.000000   
proxy                     6509.0      17303.0         6.0           0.018496   
mtproto                    388.0       1560.0         6.0           0.001103   
binary-ninja                 0.0       1529.0         5.0           0.000000   
blender                   1398.0      14835.0         5.0           0.003973   
reverse-engineering       4854.0      27503.0         5.0           0.013793   
machine-learning         13924.0     273116.0         5.0           0.039567   
pwn                          0.0         13.0         5.0           0.000000   
debugging                   12.0       3646.0         5.0           0.000034   
addon                      396.0       3702.0         5.0           0.001125   
gdb                          0.0       1789.0         5.0           0.000000   
exploit-development          0.0         37.0         5.0           0.000000   
ctf                        538.0       4611.0         5.0           0.001529   
exploit                    538.0       1092.0         5.0           0.001529   
malware-analysis           748.0       6526.0         5.0           0.002126   
blender-addon                0.0       2455.0         5.0           0.000000   
ida-pro                    186.0       2799.0         5.0           0.000529   
manuelbastionilab            0.0          0.0         4.0           0.000000   
tensorflow                6864.0     118520.0         3.0           0.019505   
elixir                       0.0          0.0         3.0           0.000000   
devops                    1667.0      27500.0         3.0           0.004737   
formatter                    0.0       1095.0         3.0           0.000000   
polyglot                     0.0       1114.0         3.0           0.000000   
game                      1689.0      30034.0         3.0           0.004799   

                     valid topic ratio  test topic ratio  
topic name                                                
python                        0.394271          0.300000  
docker                        0.013184          0.058824  
linux                         0.018949          0.047059  
hacktoberfest                 0.071546          0.047059  
telegram                      0.002118          0.035294  
mtproxy                       0.000000          0.035294  
async-mtproto-proxy           0.000000          0.035294  
proxy                         0.003190          0.035294  
mtproto                       0.000288          0.035294  
binary-ninja                  0.000282          0.029412  
blender                       0.002735          0.029412  
reverse-engineering           0.005071          0.029412  
machine-learning              0.050354          0.029412  
pwn                           0.000002          0.029412  
debugging                     0.000672          0.029412  
addon                         0.000683          0.029412  
gdb                           0.000330          0.029412  
exploit-development           0.000007          0.029412  
ctf                           0.000850          0.029412  
exploit                       0.000201          0.029412  
malware-analysis              0.001203          0.029412  
blender-addon                 0.000453          0.029412  
ida-pro                       0.000516      

В тестовой выборке хорошо представлены тематики:

docker	

linux

# Сравнение датасетов с ошибками и датасетов с корректным кодом

## На уровне репозиториев

Сравним сколько у обучающих подвыборок из каждого класса имеется общих репозиториев

In [162]:
modified_train_repos = [repo_name.replace('/', '_') for repo_name in train_repos]
modified_valid_repos = [repo_name.replace('/', '_') for repo_name in valid_repos]
modified_test_repos = [repo_name.replace('/', '_') for repo_name in test_repos]

In [149]:
len(set(modified_train_repos) & set(train_repos_)), min(len(modified_train_repos), len(train_repos_))

(94, 562)

Таким образом, имеется 94 репозитория, сниппеты из которых представлены как обучающей выборке кода с ошибками, так и в обучающей выборке с корректным кодом. В обучающей выборке с кодом с ошибками представлено 562 репозитория. Таким образом, 16% репозиториев, представлены в обеих выборках.

Теперь для валидационной выборки:

In [146]:
len(set(modified_valid_repos) & set(valid_repos_)), min(len(modified_valid_repos), len(valid_repos_))

(67, 68)

Таким образом, имеется 67 репозитория, сниппеты из которых представлены как валидационной выборке кода с ошибками, так и в валидационной выборке с корректным кодом. В валидационной выборке с кодом с ошибками представлено 68 репозиториев. Таким образом, почти 100% репозиториев, представлены в обеих выборках.

Теперь для тестовой выборки:

In [147]:
len(set(modified_test_repos) & set(test_repos_)), min(len(modified_test_repos), len(test_repos_))

(1, 68)

Таким образом, имеется 1 репозиторий, сниппеты из которых представлены как тестовой выборке кода с ошибками, так и в тестовой выборке с корректным кодом. В тестовой выборке с кодом с ошибками представлено 68 репозиториев. Таким образом, только 1.5% репозиториев представлены в обеих выборках.

## На уровне сниппетов

Теперь подсчитаем сколько сниппетов лежит в общих репозиториях в обучающих подвыборках.

In [172]:
source_code_data_train.loc[source_code_data_train['repo_name'].apply(lambda x: x.replace('/', '_') in (set(modified_train_repos) & set(train_repos_)))].shape[0], source_code_data_train.shape[0]

(2302, 14118)

Таким образом, 2302 сниппета кода из 14118 сниппетов обучающей выборки кода с ошибками имеют репозитории, общие для обеих подвыборок корректного кода и кода с ошибками.

In [171]:
source_code_data_train_.loc[source_code_data_train_['repo_name'].apply(lambda x: x in (set(modified_train_repos) & set(train_repos_)))].shape[0], source_code_data_train_.shape[0]

(54104, 351912)

Таким образом, 54104 сниппета кода из 351912 сниппетов обучающей выборки корректного кода имеют репозитории, общие для обеих подвыборок корректного кода и кода с ошибками.

Теперь подсчитаем сколько сниппетов лежит в общих репозиториях в валидационных подвыборках.

In [173]:
source_code_data_valid.loc[source_code_data_valid['repo_name'].apply(lambda x: x.replace('/', '_') in (set(modified_valid_repos) & set(valid_repos_)))].shape[0], source_code_data_valid.shape[0]

(9407, 9457)

Таким образом, 9407 сниппета кода из 9457 сниппетов валидационной выборки кода с ошибками имеют репозитории, общие для обеих подвыборок корректного кода и кода с ошибками.

In [174]:
source_code_data_valid_.loc[source_code_data_valid_['repo_name'].apply(lambda x: x in (set(modified_valid_repos) & set(valid_repos_)))].shape[0], source_code_data_valid_.shape[0]

KeyboardInterrupt: 

Таким образом, подавляющее большинство сниппетов валидационной выборки корректного кода имеют репозитории, общие для обеих подвыборок корректного кода и кода с ошибками.

Теперь подсчитаем сколько сниппетов лежит в общих репозиториях в тестовых подвыборках.

In [175]:
source_code_data_test.loc[source_code_data_test['repo_name'].apply(lambda x: x.replace('/', '_') in set(modified_test_repos) & set(test_repos_))].shape[0], source_code_data_test.shape[0]

(1, 161)

Таким образом, 1 сниппет кода из 161 сниппетов тестовой выборки кода с ошибками имеют репозитории, общие для обеих подвыборок корректного кода и кода с ошибками.

In [176]:
source_code_data_test_.loc[source_code_data_test_['repo_name'].apply(lambda x: x in set(modified_test_repos) & set(test_repos_))].shape[0], source_code_data_test_.shape[0]

(1, 170)

Таким образом, 1 сниппет кода из 170 сниппетов тестовой выборки корректного кода имеют репозитории, общие для обеих подвыборок корректного кода и кода с ошибками.

## На уровне топиков

In [178]:
all_topics_counts = pd.concat((topics_count, topics_count_), axis=1)

In [179]:
all_topics_counts

train count  valid count  test count  train topic ratio  \
topic name                                                                     
python                    9538.0       7162.0       131.0           0.675591   
machine-learning          2239.0        661.0        10.0           0.158592   
hacktoberfest             2098.0       2365.0        34.0           0.148605   
pandas                    1733.0        297.0         5.0           0.122751   
numpy                     1542.0        210.0         3.0           0.109222   
...                          ...          ...         ...                ...   
programmable-engine          NaN          NaN         NaN                NaN   
speech-corpora               NaN          NaN         NaN                NaN   
cmu-sphinx                   NaN          NaN         NaN                NaN   
sequitur                     NaN          NaN         NaN                NaN   
voxforge                     NaN          NaN         NaN                NaN   

                     valid topic ratio  test topic ratio  train count  \
topic name                                                              
python                        0.757323          0.813665     106168.0   
machine-learning              0.069895          0.062112      13924.0   
hacktoberfest                 0.250079          0.211180      16137.0   
pandas                        0.031405          0.031056       3348.0   
numpy                         0.022206          0.018634        867.0   
...                                ...               ...          ...   
programmable-engine                NaN               NaN          0.0   
speech-corpora                     NaN               NaN          0.0   
cmu-sphinx                         NaN               NaN          0.0   
sequitur                           NaN               NaN          0.0   
voxforge                           NaN               NaN          0.0   

                     valid count  test count  train topic ratio  \
topic name                                                        
python                 2138474.0        51.0           0.301689   
machine-learning        273116.0         5.0           0.039567   
hacktoberfest           388057.0         8.0           0.045855   
pandas                   48091.0         0.0           0.009514   
numpy                    64166.0         0.0           0.002464   
...                          ...         ...                ...   
programmable-engine          0.0         1.0           0.000000   
speech-corpora               0.0         1.0           0.000000   
cmu-sphinx                   0.0         1.0           0.000000   
sequitur                     0.0         1.0           0.000000   
voxforge                     0.0         1.0           0.000000   

                     valid topic ratio  test topic ratio  
topic name                                                
python                        0.394271          0.300000  
machine-learning              0.050354          0.029412  
hacktoberfest                 0.071546          0.047059  
pandas                        0.008867          0.000000  
numpy                         0.011830          0.000000  
...                                ...               ...  
programmable-engine           0.000000          0.005882  
speech-corpora                0.000000          0.005882  
cmu-sphinx                    0.000000          0.005882  
sequitur                      0.000000          0.005882  
voxforge                      0.000000          0.005882  

[14176 rows x 12 columns]